confection/tests/test_frozen_structures.py

```diff
@@ -30,3 +34,30 @@ def test_frozen_dict():
 
     with raises(NotImplementedError, match="frozen dictionary"):
         frozen[10] = 1
+
+
+@pytest.mark.parametrize("frozen_type", ("dict", "list"))
+def test_frozen_struct_deepcopy(frozen_type):
+    """Test whether setting default values for a FrozenDict/FrozenList works within a config, which utilizes
+    deepcopy."""
+    registry.bar = catalogue.create("confection", "bar", entry_points=False)
+
+    @registry.bar.register("foo_dict.v1")
+    def make_dict(values: Dict[str, int] = SimpleFrozenDict(x=3)):
+        return values
+
+    @registry.bar.register("foo_list.v1")
+    def make_list(values: List[int] = SimpleFrozenList([1, 2, 3])):
+        return values
+
+    cfg = Config()
+    resolved = registry.resolve(
+        cfg.from_str(
+            f"""
+            [something]
+            @bar = "foo_{frozen_type}.v1"        
+            """
+        )
+    )
+
+    assert isinstance(resolved["something"], Dict if frozen_type == "dict" else List)
```

Comment:
Why can't this test that it's specifically an instance of `SimpleFrozenDict` or `SimpleFrozenList`? It looks like in the case of `"dict"`, it really is just a Python `dict` and not a `SimpleFrozenDict` and so this test would fail with the below edit, but why?

```suggestion
    assert isinstance(resolved["something"], SimpleFrozenDict if frozen_type == "dict" else SimpleFrozenList)
```

Reply:
To be honest, I'm not sure - I was surprised myself.

Reply:
Should we look further into this?

Reply:
If you had noticed this when implementing the test, it would have been good to already call it out when creating the PR, because it's the kind of thing that can be overlooked when reviewing.

I do think we should spent a bit more time investigating why this happens. It feels like it might be because of a slightly different bug that's interfering here.

Reply:
> If you had noticed this when implementing the test, it would have been good to already call it out when creating the PR, because it's the kind of thing that can be overlooked when reviewing.

Fair point.

> I do think we should spent a bit more time investigating why this happens. It feels like it might be because of a slightly different bug that's interfering here.

Will do.

Reply:
This is strange. This doesn't seem to be related to `confection`...the `deepcopy` mechanism _returns_ a `SimpleFrozenDict`, but:
```python
    @registry.bar.register("foo_dict.v1")
    def make_dict(values: Dict[str, int] = SimpleFrozenDict()):
        x = isinstance(values, SimpleFrozenDict)  # x is False
        return values
```
It's not clear to me yet why we don't have a `SimpleFrozenDict` at this point.

Reply:
Functionally that shouldn't be an issue though, I think :thinking: What's your take? @svlandeg

Reply:
Strange. Let's jot it down in the backlog to look into at a later point, so we can continue with this PR as is?